File: _labs/lab05/cicd/R/penguinpredictor/R/logging_setup.R
#' Configure logging with file and console output
#'
#' Creates the log directory if it does not exist and returns the path
#' to the log file.
#'
#' @param log_dir character path to log directory
#' @param level log threshold level (default logger::INFO)
#' @return character path to log file (invisibly)
#' @export
setup_logging <- function(log_dir = "logs", level = logger::INFO) {
  if (!dir.exists(log_dir)) {
    dir.create(log_dir, recursive = TRUE)
  }
  log_file <- file.path(log_dir, "shiny_app.log")

  logger::log_appender(logger::appender_tee(log_file))
  logger::log_threshold(level)
  logger::log_layout(logger::layout_glue_generator(
    format = "{time} - {level} - {msg}"
  ))

  invisible(log_file)
}

File: _labs/lab05/cicd/Python/penguin_predictor/penguin_predictor/logging_setup.py
import logging
from pathlib import Path


def setup_logging(log_dir="logs", level=logging.INFO):
    """Configure logging with file and console output.

    Creates the log directory if it does not exist and returns the path
    to the log file.
    """
    log_path = Path(log_dir)
    log_path.mkdir(exist_ok=True)
    log_file = log_path / "shiny_app.log"

    logging.basicConfig(
        level=level,
        format="%(asctime)s - %(levelname)s - %(message)s",
        handlers=[
            logging.FileHandler(log_file, mode="a"),
            logging.StreamHandler(),
        ],
    )
    return str(log_file)
